### ISSUE TITLE
processing of the outputs

### ISSUE THREAD

**Issue by lyupei** [MandyMo/pytorch_HMR#3]
Hi, it's me again.

So far I can train the model using your scripts, but still have some problem processing the outputs.

I wrote a script almost the same as the original demo.py, replacing the tensorflow inference code with forward functions in model.py, and processed the 2d kps (output range is (-1,1)) to be drawed.

But I can't get the right output rendering image.

In fact, even when I run with the same weights file twice, I can't get the same outputs.

I'm new to pytorch, maybe I have encountered some Novice problem。

Ps: I use the copy_state_dict funtion to load weights file.

**Comment by lyupei** [comment]
I found that I forgot to discard the dropout layer during inference. Now I can get the same outputs.

But the result isn't right still.

Maybe I should try more iterations?

**Comment by MandyMo** [comment]
I train the network with 2 TitanXp GPU，with 1000000 iterations.(encoder is resnet50)

Finally, when testing, you do some preprocess as in the dataloader.

**Comment by MandyMo** [comment]
``    for (batch_index, samples) in enumerate(data_loader):
        images, paths = samples['image'].float().cuda(), samples['image_name']
        (theta, verts, j2ds, j3d, Rs) = model(images)[-1]

        batch_size = images.shape[0]
        for idx in range(batch_size):
            image = ((images[idx].cpu() + 1) * 0.5 * 255).numpy().copy()
            image = image.astype(np.uint8).transpose((1, 2, 0)).copy()
            j2d = j2ds[idx].detach().cpu().numpy()
            #j2d[:, 1] = -j2d[:, 1]
            j2d = (j2d + 1) * 112
            j2d = np.concatenate((j2d, one), axis = 1)[:14]
            draw_lsp_14kp__bone(image, j2d)

            dst_image = np.zeros((224, 448, 3))
            dst_image[:,224:,:] = image[:,:,:]


            j2d = samples['kp_2d'][idx].cpu().numpy()
            j2d = (j2d + 1) * 112
            image = ((images[idx].cpu() + 1) * 0.5 * 255).numpy().copy()
            image = image.astype(np.uint8).transpose((1, 2, 0)).copy() 
            draw_lsp_14kp__bone(image, j2d)
            dst_image[:,:224,:] = image[:,:,:]
            

            cv2.imshow('window', cv2.resize(dst_image.astype(np.uint8), (1024, 512), interpolation = cv2.INTER_CUBIC))
            cv2.waitKey(0)

            image_name = paths[idx]
            short_name = os.path.basename(image_name)
            vert = verts[idx].detach().cpu().numpy()
            save_name = os.path.join('E:/objs', short_name + '.obj')
            smpl.save_obj(vert, save_name)

**Comment by MandyMo** [comment]
def _test_images(model, src_image_folder, save_folder, crop_size = 224):
    if not os.path.exists(save_folder):
        os.makedirs(save_folder)

    smpl = model.smpl
    w_visible = np.ones([19, 1], dtype = np.float)

    for image_path in glob.glob(os.path.join(src_image_folder, '*.png')):
        src_image = cv2.imread(image_path)
        dst_iamge = src_image.copy()
        dst_iamge = cv2.resize(dst_iamge, (crop_size, crop_size), interpolation = cv2.INTER_CUBIC).transpose((2, 0, 1))
        input_image = ((torch.from_numpy(dst_iamge).float().unsqueeze(0) / 255) - 0.5).cuda() * 2
        (theta, verts, j2d, j3d, Rs) = model(input_image)[-1]

        vert = verts[0].detach().cpu().numpy()
        j2d = j2d[0].detach().cpu().numpy()
        j2d = (j2d + 1) * 500
        j2d = np.concatenate((j2d, w_visible), axis = 1)

        draw_lsp_14kp__bone(src_image, j2d[:14,:])
    
        save_obj_name = os.path.basename(image_path).split('.')[0] + '.obj'
        smpl.save_obj(vert, os.path.join(save_folder, save_obj_name))
        save_image_name = os.path.basename(image_path)
        cv2.imwrite(os.path.join(save_folder, save_image_name), src_image)


this is my testing code.

**Comment by MandyMo** [comment]
`
def _test_images(model, src_image_folder, save_folder, crop_size = 224):
    if not os.path.exists(save_folder):
        os.makedirs(save_folder)

    smpl = model.smpl
    w_visible = np.ones([19, 1], dtype = np.float)

    for image_path in glob.glob(os.path.join(src_image_folder, '*.png')):
        src_image = cv2.imread(image_path)
        dst_iamge = src_image.copy()
        dst_iamge = cv2.resize(dst_iamge, (crop_size, crop_size), interpolation = cv2.INTER_CUBIC).transpose((2, 0, 1))
        input_image = ((torch.from_numpy(dst_iamge).float().unsqueeze(0) / 255) - 0.5).cuda() * 2
        (theta, verts, j2d, j3d, Rs) = model(input_image)[-1]

        vert = verts[0].detach().cpu().numpy()
        j2d = j2d[0].detach().cpu().numpy()
        j2d = (j2d + 1) * 500
        j2d = np.concatenate((j2d, w_visible), axis = 1)

        draw_lsp_14kp__bone(src_image, j2d[:14,:])
    
        save_obj_name = os.path.basename(image_path).split('.')[0] + '.obj'
        smpl.save_obj(vert, os.path.join(save_folder, save_obj_name))
        save_image_name = os.path.basename(image_path)
        cv2.imwrite(os.path.join(save_folder, save_image_name), src_image)
`

**Comment by lyupei** [comment]
Hi, I was on a vacation so sorry for the late reply.
I trained this model from sctrach for 500000 iterations:
OrderedDict([('time', '2016-02-14-08:01:44'), ('iter', 500000), ('e_loss', 18.80540905892849), ('2d_loss', 3.607551097869873), ('3d_loss', 1.0080459117889404), ('shape_loss', 0.1922520250082016), ('pose_loss', 6.980734348297119), ('e_disc_loss', 7.0168256759643555), ('d_disc_loss', 12.02401351928711), ('d_disc_real', 6.099047660827637), ('d_disc_predict', 5.924966335296631), ('title', '500000_')])

But I still can't get the right output. Here is my demo.py:

    model = HMRNetBase().cuda()
    w_visible = np.ones([19, 1], dtype=np.float)
    copy_state_dict(
        model.state_dict(),
        torch.load("../../500000_12.271408440545201_generator.pkl"),
        prefix='module.'
    )
    smpl = model.smpl
    src_image = cv2.imread(img_path)
    dst_image = src_image.copy()
    dst_image = dst_image.transpose(2, 0, 1)
    input_img = ((torch.from_numpy(dst_image).float().unsqueeze(0) /255) - 0.5).cuda() * 2
    theta, verts, j2d, j3d, RS = model.forward(input_img)[-1]
    vert = verts[0].detach().cpu().numpy()
    j2d = j2d[0].detach().cpu().numpy()
    j2d = (j2d+1)*112
    j2d = np.concatenate((j2d,  w_visible),axis = 1)
    draw_lsp_14kp__bone(src_image, j2d[:14,:])

    smpl.save_obj(vert,"./save.obj")
    cv2.imwrite("result.jpg",src_image)

I used the demo images offered by hmr so crop is not needed. And I discarded the dropout layer,
Is there anything wrong with my processing?
Looking forward for your help!

**Comment by lyupei** [comment]
I modified the code to save and view the intermediate result using tensorboardX, and found that the draw_lsp_14kp_bone() result seems right during training.
So I rewrite the inference.py the seem as trainer.py, and found that if I set batch_size(of dataloader)=1, the result differs very few on different input images. But if I set batch_size>1, the result looks right.
Is this working properly?

**Comment by lyupei** [comment]
hmm.....
I forgot set the generator.Train(False), now it works fine!

**Comment by weigq** [comment]
@lyupei 
Not works here. Can you provide some inference samples? like the input images and the corresponding mesh?
Thx.

**Comment by tszhang97** [comment]
@lyupei  Hi, could share you modified version of demo.py? I met some problems of test on pretrained model.
